contracts/bonding/BondingManager.sol

```diff
@@ -35,10 +35,12 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         uint256 rewardCut;                                              // % of reward paid to transcoder by a delegator
         uint256 feeShare;                                               // % of fees paid to delegators by transcoder
         uint256 pricePerSegmentDEPRECATED;                              // DEPRECATED - DO NOT USE
-        uint256 pendingRewardCut;                                       // Pending reward cut for next round if the transcoder is active
-        uint256 pendingFeeShare;                                        // Pending fee share for next round if the transcoder is active
+        uint256 pendingRewardCutDEPRECATED;                             // DEPRECATED - DO NOT USE
+        uint256 pendingFeeShareDEPRECATED;                              // DEPRECATED - DO NOT USE
         uint256 pendingPricePerSegmentDEPRECATED;                       // DEPRECATED - DO NOT USE
         mapping (uint256 => EarningsPool.Data) earningsPoolPerRound;    // Mapping of round => earnings pool for the round
+        uint256 lastActiveStakeUpdateRound;                             // Round for which the stake was last updated while the transcoder is active
+        uint256 activationRound;                                        // Round in which the transcoder became active - 0 if inactive
```

Comment:
I think we might need to add a `deactivationRound` field. The reasoning is the following:

If we're establishing that the active set and the total active stake is locked in for a round then if a transcoder resigns or is evicted in round N, it should exit the active set and be deactivated in round N + 1. However, given the current definition of `isActiveTranscoder()`, a transcoder would be considered inactive in round N as soon as it resigns or is evicted. If we add a `deactivationRound` field then we can change the definition of `isActiveTranscoder()` to:

```
return t.activationRound != 0 && t.activationRound <= currentRound && currentRound < t.deactivationRound
```

The default value for `t.deactivationRound` can be 0 and when a transcoder is activated we can set `t.deactivationRound` to some very large integer. When a transcoder is resigned or is evicted, we set `t.deactivationRound = currentRound + 1`.

Reply:
Hmm I actually thought it would make sense for the deactivation of a transcoder to happen immediately and not wait until the next round.

1. A transcoder fully unbonds, and thereby also becomes unregistered so it shouldn't be marked as active either I think. 

2. A transcoder gets slashed (which won't occur in SF), should a misbehaving transcoder be kept in the active set until the end of the round, to me that seems contradictory?

Reply:
Some additional reasons for deactivating a transcoder in the next round:

- If an active transcoder is evicted in round N prior to calling `reward()`, then that transcoder will never be able to call `reward()` for that round since we require that only active transcoders in the current round can call `reward()`. This would run counter to the design principle that the active transcoder set and the total active stake is locked in at the beginning of a round after the round is initialized.
- Reduces active set churn in a given round. Consider the scenario where N active transcoders are evicted because N new transcoders with more stake join. If we immediately deactivate the old transcoders and the new transcoders are not activated until the next round then in the current round there are N less active transcoders. As a result, from a broadcaster's view, the # of eligible service providers on the network in the current round just dropped by N. If we instead deactivate transcoders in the next round then we maintain a full active set for the duration of the round and when the next round begins we "rotate" in any new transcoders that are replacing the deactivated transcoders.

> A transcoder fully unbonds, and thereby also becomes unregistered so it shouldn't be marked as active either I think.

Active transcoders that are pending deactivation in the next round can continue doing work on the network up until the start of the next round and then they can go offline permanently. Note that active transcoders that willfully resign in a round can go offline immediately anyway and there is no mechanism to force them to continue working until the beginning of the next round. But, they have the option to do so and to continue collecting fees until they become deactivated in the next round.

> A transcoder gets slashed (which won't occur in SF), should a misbehaving transcoder be kept in the active set until the end of the round, to me that seems contradictory?

In this case, we can keep the transcoder in the active set for the remainder of the round, but clients could watch for slashing events and just not work with an active transcoder that has been slashed before. This is the one situation where the effective active set available to a broadcaster would definitely be reduced since broadcasters would refuse to work with slashed active transcoders that are pending deactivation, but no replacement transcoders will be activated until the next round.

Reply:
`return t.activationRound != 0 && t.activationRound <= currentRound && currentRound < t.deactivationRound`

Since we currently zero `t.activationRound` when resigning a transcoder, then  this will still return `false` unless we don't zero `t.activationRound` , right?

Reply:
Yeah we can not zero `t.activationRound` when a transcoder is pending deactivation and instead:

Let `MAX_FUTURE_ROUND = 2^256 - 1`.

When a transcoder is pending activation next round i.e. it joins the pool:

- Set `activationRound = currentRound + 1`
- Set `deactivationRound = MAX_FUTURE_ROUND`

When a transcoder is pending deactivation in the next round i.e. it leaves the pool:

- Set `deactivationRound = currentRound + 1`

Reply:
added in fe63b1d